src/app.js

```diff
@@ -502,4 +290,76 @@ class SAFEApp extends EventEmitter {
 
 SAFEApp.logFilename = null;
 
-module.exports = SAFEApp;
+/**
+* Init logging on the underlying library only if it wasn't done already
+*/
+const initLogging = (appInfo, options) => {
+  if (options.log && !SAFEApp.logFilePath) {
+    let filename = `${appInfo.name}.${appInfo.vendor}`.replace(/[^\w\d_\-.]/g, '_');
+    filename = `${filename}.log`;
+    return lib.app_init_logging(filename)
+      .then(() => lib.app_output_log_path(filename))
+      .then((logPath) => { SAFEApp.logFilePath = logPath; })
+      .catch((err) => {
+        throw makeError(errConst.LOGGER_INIT_ERR.code, errConst.LOGGER_INIT_ERR.msg(err));
+      });
+  }
+};
+
+/**
+* Set additional search path for the config files if it was requested in
+* the options. E.g. log.toml and crust.config files will be search
+* in this additional search path.
+*/
+const setSearchPath = (options) => {
+  if (options.configPath) {
+    return lib.app_set_additional_search_path(options.configPath)
+      .catch((err) => {
+        throw makeError(errConst.CONFIG_PATH_ERROR.code, errConst.CONFIG_PATH_ERROR.msg(err));
+      });
+  }
+};
+
+/**
+* Validates appInfo and properly handles error
+*/
+const validateAppInfo = (_appInfo) => {
+  const appInfo = _appInfo;
+  const appInfoMustHaveProperties = ['id', 'name', 'vendor'];
+  let bool = false;
+  const hasCorrectProperties = appInfoMustHaveProperties.every((prop) => {
+    if (appInfo && appInfo[prop]) {
+      appInfo[prop] = appInfo[prop].trim();
+      bool = Object.prototype.hasOwnProperty.call(appInfo, prop) && appInfo[prop];
+    }
+
+    return bool;
+  });
+
+  if (!hasCorrectProperties) {
+    throw makeError(errConst.MALFORMED_APP_INFO.code, errConst.MALFORMED_APP_INFO.msg);
+  }
+};
+
+/**
+ * Primary exported module handles async functions first, \
+ * handles async error, \
+ * returns instance of SAFEApp.
+ */
+async function asyncSAFEApp(appInfo, networkStateCallBack, options) {
```

Comment:
Let's keep a consistency with using arrow functions.
Perhaps we should just expose the SAFEApp constructor and make this function to be like an `init`/`initialise` async class method instead of having it as a factory function. in this way we could keep the abstraction of instantiating and initialising the SAFEApp instance all within the `initialiseApp` API function as is, the only difference is that if the user creates the SAFEApp instance it will need to call `init/`initalise` afterwards when it wasn't needed before.

Reply:
Aye, this is a good idea and it wouldn't being changing the exposed API, we'd just be calling `app.init(...)` inside of `initialiseApp`. Ok, I'll change this around.